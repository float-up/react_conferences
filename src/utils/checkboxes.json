[
  {
    "title": "RU",
    "type": "lang"
  },
  {
    "title": "EN",
    "type": "lang"
  },
  {
    "title": "HOT",
    "type": "level"
  },
  {
    "title": "INTERMEDIATE",
    "type": "level"
  },
  {
    "title": "ADVANCED",
    "type": "level"
  },
  {
    "title": "HARDCORE",
    "type": "level"
  },
  {
    "title": "ACADEMIC",
    "type": "level"
  }
]
