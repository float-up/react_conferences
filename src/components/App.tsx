import React, {useContext, useEffect} from 'react';
import ConfsStore from "../store/ConfsStore";
import {ConferenceCard} from './ConferenceCard';
import Input from "./Input";
import {Checkbox} from "./Checkbox";
import {Button} from "./Button";
import {Loader} from "./Loader";
import {observer} from "mobx-react-lite";
import checkboxes from '../utils/checkboxes.json';

const App = () => {

    const confsStore = useContext(ConfsStore);
    const {
        confs,
        loading,
        isNotFilers,
        checkboxesLang,
        checkboxesLevel,
        setFilterInput,
        toggleCheckboxLevel,
        toggleCheckboxLang,
        resetFilters,
        getConfs,
    } = confsStore;

    useEffect(() => {
        getConfs();
    }, [getConfs])

    return (
        <div className="App">
            <nav>
                <div className="nav-wrapper orange lighten-3">
                    <a href="https://jug.ru" className="brand-logo center">JUG.RU Confs</a>
                </div>
            </nav>
            <div className="container">

                <div className="checkboxes">
                    {checkboxes.map((cb, index) =>
                        <Checkbox
                            text={cb.title}
                            onChange={cb.type === 'lang' ? toggleCheckboxLang : toggleCheckboxLevel}
                            key={index}
                            checked={[...checkboxesLevel, ...checkboxesLang].includes(cb.title)}
                        />)
                    }
                </div>

                <div className="row">
                    <div className="col s10">
                        <Input placeholder="Search" onChange={setFilterInput}/>
                    </div>
                    <div className="col s2">
                        <Button text="Reset Filters" onClick={resetFilters} disable={isNotFilers}/>
                    </div>
                </div>
                <div className="row">
                    {loading ? <Loader /> : confs.length ?
                        confs.map(conf => <div className="col s6" key={conf.id}><ConferenceCard confInfo={conf}/></div>) :
                        <h4>Nothing to show</h4>
                    }
                </div>
            </div>
        </div>
    );
}

export default observer(App);
