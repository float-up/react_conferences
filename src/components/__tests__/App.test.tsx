import React, {useContext} from "react"
import {mount, render, shallow} from 'enzyme'
import checkboxes from '../../utils/checkboxes.json';

import App from "../App";

const onclick = jest.fn();
jest.mock('../../store/ConfsStore')

describe("JUG.RU intergrate testing", () => {

    let wrapper = mount(<App/>)

    beforeEach(() => {
        wrapper
            .find('button')
            .simulate('click', {target: {}})
    })

    it('correct rendering App', () => {
        expect(wrapper).toMatchSnapshot();
    })


    it('correct render checkboxes', () => {

        expect(wrapper.find('.checkbox')).toHaveLength(checkboxes.length);
    })


    it('correct render Input', () => {

        expect(wrapper.find('.inputFilter')).toHaveLength(1);
    })


    it('correct render Button reset filters', () => {

        expect(wrapper.find('button').text()).toEqual('Reset Filters');
    })

    it('correct Button reset is disabled', () => {
        expect(wrapper.find('button').text()).toEqual('Reset Filters');
        expect(wrapper.find('button').prop('disabled')).toBe(true)
    })

    it('correct Button reset is not disabled', () => {
        expect(wrapper.find('button').text()).toEqual('Reset Filters');
        wrapper
            .find('input[type="checkbox"]')
            .at(0)
            .simulate('change', {target: {checked: true}})

        expect(wrapper.find('button').prop('disabled')).toBe(false)
    })

    it('correct render Conference Cards', () => {
        expect(wrapper.find('.card')).toHaveLength(7);
    })


    it('check filter working #1 - toggle checkboxes: RU', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(0)
            .simulate('change', {target: {checked: true}})

        expect(wrapper.find('.card')).toHaveLength(4);
        expect(wrapper.find('.card-level').at(0).text()).toMatch('RU')
        expect(wrapper.find('.card-level').at(1).text()).toMatch('RU')
        expect(wrapper.find('.card-level').at(2).text()).toMatch('RU')
        expect(wrapper.find('.card-level').at(3).text()).toMatch('RU')
    })


    it('check filter working #2 - toggle checkboxes: RU and HARDCORE', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(0)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="checkbox"]')
            .at(5)
            .simulate('change', {target: {checked: true}})

        expect(wrapper.find('.card')).toHaveLength(2);
        expect(wrapper.find('.card-level').at(0).text()).toEqual('HARDCORE / RU')
        expect(wrapper.find('.card-level').at(1).text()).toEqual('HARDCORE / RU')
    })


    it('check filter working #3 - toggle checkboxes: EN and ADVANCED and ACADEMIC', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(1)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="checkbox"]')
            .at(4)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="checkbox"]')
            .at(6)
            .simulate('change', {target: {checked: true}})

        expect(wrapper.find('.card')).toHaveLength(2);
        expect(wrapper.find('.card-level').at(0).text()).toEqual('ADVANCED / EN')
        expect(wrapper.find('.card-level').at(1).text()).toEqual('ACADEMIC / EN')
    })


    it('check filter working #4 - toggle checkboxes: HARDCODE and INTERMEDIATE', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(3)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="checkbox"]')
            .at(5)
            .simulate('change', {target: {checked: true}})


        expect(wrapper.find('.card')).toHaveLength(3);
        expect(wrapper.find('.card-level').at(0).text()).toEqual('INTERMEDIATE / EN')
        expect(wrapper.find('.card-level').at(1).text()).toEqual('HARDCORE / RU')
        expect(wrapper.find('.card-level').at(2).text()).toEqual('HARDCORE / RU')
    })


    it('check filter working #5 - toggle checkboxes: RU and INTERMEDIATE', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(0)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="checkbox"]')
            .at(3)
            .simulate('change', {target: {checked: true}})


        expect(wrapper.find('.card')).toHaveLength(0);
    })


    it('check filter working #6 - filter input value: "Дмитрий"', () => {


        wrapper
            .find('input[type="text"]')
            .at(0)
            .simulate('change', {target: {value: 'Дмитрий'}})

        expect(wrapper.find('.card')).toHaveLength(2);

        expect(wrapper.find('.card-lector').at(0).text()).toEqual('Дмитрий Пацура')
        expect(wrapper.find('.card-lector').at(1).text()).toEqual('Дмитрий Волошин')
    })


    it('check filter working #7 - toggle checkboxes: RU and HARDCORE and filter input value: "Ом"', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(0)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="checkbox"]')
            .at(5)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="text"]')
            .at(0)
            .simulate('change', {target: {value: 'Ом'}})

        expect(wrapper.find('.card')).toHaveLength(2);

        expect(wrapper.find('.card-title').at(0).text()).toEqual('Разработка компилятора для TypeScript на TypeScript на базе LLVM')
        expect(wrapper.find('.card-lector').at(1).text()).toEqual('Роман Дворнов')
    })


    it('check filter working #8 - toggle checkboxes: RU and filter input value: "ли" and click Reset button', () => {

        wrapper
            .find('input[type="checkbox"]')
            .at(0)
            .simulate('change', {target: {checked: true}})

        wrapper
            .find('input[type="text"]')
            .at(0)
            .simulate('change', {target: {value: 'ли'}})

        expect(wrapper.find('.card')).toHaveLength(2);

        expect(wrapper.find('.card-title').at(0).text()).toEqual('GraphQL-фрагменты на клиенте: История появления, ошибки использования')
        expect(wrapper.find('.card-title').at(1).text()).toEqual('Может ли компьютер молиться на благо всех живых существ? Молитва на JavaScript и WebGL')

        wrapper
            .find('button')
            .simulate('click', {target: {}})

        expect(wrapper.find('.card')).toHaveLength(7);
    })

})
