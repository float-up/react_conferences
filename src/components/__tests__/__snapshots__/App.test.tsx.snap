// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JUG.RU intergrate testing correct rendering App 1`] = `
<App>
  <div
    className="App"
  >
    <nav>
      <div
        className="nav-wrapper orange lighten-3"
      >
        <a
          className="brand-logo center"
          href="https://jug.ru"
        >
          JUG.RU Confs
        </a>
      </div>
    </nav>
    <div
      className="container"
    >
      <div
        className="checkboxes"
      >
        <Checkbox
          checked={false}
          key="0"
          onChange={[Function]}
          text="RU"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              RU
            </span>
          </label>
        </Checkbox>
        <Checkbox
          checked={false}
          key="1"
          onChange={[Function]}
          text="EN"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              EN
            </span>
          </label>
        </Checkbox>
        <Checkbox
          checked={false}
          key="2"
          onChange={[Function]}
          text="HOT"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              HOT
            </span>
          </label>
        </Checkbox>
        <Checkbox
          checked={false}
          key="3"
          onChange={[Function]}
          text="INTERMEDIATE"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              INTERMEDIATE
            </span>
          </label>
        </Checkbox>
        <Checkbox
          checked={false}
          key="4"
          onChange={[Function]}
          text="ADVANCED"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              ADVANCED
            </span>
          </label>
        </Checkbox>
        <Checkbox
          checked={false}
          key="5"
          onChange={[Function]}
          text="HARDCORE"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              HARDCORE
            </span>
          </label>
        </Checkbox>
        <Checkbox
          checked={false}
          key="6"
          onChange={[Function]}
          text="ACADEMIC"
        >
          <label
            className="checkbox"
          >
            <input
              checked={false}
              id="indeterminate-checkbox"
              onChange={[Function]}
              type="checkbox"
            />
            <span>
              ACADEMIC
            </span>
          </label>
        </Checkbox>
      </div>
      <div
        className="row"
      >
        <div
          className="col s10"
        >
          <Input
            onChange={[Function]}
            placeholder="Search"
          >
            <div>
              <input
                className="inputFilter"
                onChange={[Function]}
                placeholder="Search"
                type="text"
                value=""
              />
            </div>
          </Input>
        </div>
        <div
          className="col s2"
        >
          <Button
            disable={true}
            onClick={[Function]}
            text="Reset Filters"
          >
            <button
              className="waves-effect waves-light btn-large"
              disabled={true}
              onClick={[Function]}
            >
              Reset Filters
            </button>
          </Button>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col s6"
          key="1"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "1",
                "lector": "Guillermo Rauch",
                "title": "Client rendering, server rendering, pre rendering. The full spectrum of JS website and application perfomance delivery",
                "type": Object {
                  "lang": "EN",
                  "level": "INTERMEDIATE",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  Client rendering, server rendering, pre rendering. The full spectrum of JS website and application perfomance delivery
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Guillermo Rauch
                </p>
                <br />
                <p
                  className="card-level"
                >
                  INTERMEDIATE
                   / 
                  EN
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
        <div
          className="col s6"
          key="2"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "2",
                "lector": "Павел Черторогов",
                "title": "GraphQL-фрагменты на клиенте: История появления, ошибки использования",
                "type": Object {
                  "lang": "RU",
                  "level": "ADVANCED",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  GraphQL-фрагменты на клиенте: История появления, ошибки использования
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Павел Черторогов
                </p>
                <br />
                <p
                  className="card-level"
                >
                  ADVANCED
                   / 
                  RU
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
        <div
          className="col s6"
          key="3"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "3",
                "lector": "Nicolo Ribaudo",
                "title": "@babel/how-to",
                "type": Object {
                  "lang": "EN",
                  "level": "ADVANCED",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  @babel/how-to
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Nicolo Ribaudo
                </p>
                <br />
                <p
                  className="card-level"
                >
                  ADVANCED
                   / 
                  EN
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
        <div
          className="col s6"
          key="4"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "4",
                "lector": "Дмитрий Пацура",
                "title": "Разработка компилятора для TypeScript на TypeScript на базе LLVM",
                "type": Object {
                  "lang": "RU",
                  "level": "HARDCORE",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  Разработка компилятора для TypeScript на TypeScript на базе LLVM
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Дмитрий Пацура
                </p>
                <br />
                <p
                  className="card-level"
                >
                  HARDCORE
                   / 
                  RU
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
        <div
          className="col s6"
          key="5"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "5",
                "lector": "Роман Дворнов",
                "title": "CSS definition syntax",
                "type": Object {
                  "lang": "RU",
                  "level": "HARDCORE",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  CSS definition syntax
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Роман Дворнов
                </p>
                <br />
                <p
                  className="card-level"
                >
                  HARDCORE
                   / 
                  RU
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
        <div
          className="col s6"
          key="6"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "6",
                "lector": "Денис Радин",
                "title": "Может ли компьютер молиться на благо всех живых существ? Молитва на JavaScript и WebGL",
                "type": Object {
                  "lang": "RU",
                  "level": "ACADEMIC",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  Может ли компьютер молиться на благо всех живых существ? Молитва на JavaScript и WebGL
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Денис Радин
                </p>
                <br />
                <p
                  className="card-level"
                >
                  ACADEMIC
                   / 
                  RU
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
        <div
          className="col s6"
          key="7"
        >
          <ConferenceCard
            confInfo={
              Object {
                "id": "7",
                "lector": "Дмитрий Волошин",
                "title": "Карьерная эффективность и карьерный путь в ИТ",
                "type": Object {
                  "lang": "EN",
                  "level": "ACADEMIC",
                },
              }
            }
          >
            <div
              className="card grey lighten-5"
            >
              <div
                className="card-content black-text"
              >
                <span
                  className="card-title"
                >
                  Карьерная эффективность и карьерный путь в ИТ
                </span>
                <br />
                <p
                  className="card-lector"
                >
                  Дмитрий Волошин
                </p>
                <br />
                <p
                  className="card-level"
                >
                  ACADEMIC
                   / 
                  EN
                </p>
              </div>
            </div>
          </ConferenceCard>
        </div>
      </div>
    </div>
  </div>
</App>
`;
